### PR TITLE
Fix Path,EMSDK and EM_CACHE not being set permanently when using --global

### PR DESCRIPTION
**Pull request by kudaba**
Going through the initial setup of emscripten the first time, the first thing I wanted to do was set it up globally to use in visual studio and found that some of the environment variables are not getting applied properly. There are still additional issues with environment variables however since EMSDK_NODE and EMSDK_PYTHON  and PATH will get installed system wide while EM_CACHE and EMSDK will only get installed to user.

FYI: I also ran into some reported issues during my initial setup that didn't block me in the end:
* pywin32 was missing, so I needed to pip install it, maybe add a not in the windows setup documentation.
* ``ERROR: The system was unable to find the specified registry key or value.`` which are not actual issues but should probably be suppressed (trying to delete registry key that doesn't exist). Happens on every activate call.
* ``(1159, 'PostMessage', 'The message can be used only with synchronous operations.')`` Only happens when system registry keys are not set, it doesn't cause them to fail though.

Edit:
So, a big update here. While the code remains to be able to set user environment variables in windows it really only runs in two modes: current session locals or system wide environment variables. I removed the code that was trying to delete the variables and causing errors since it doesn't really do anything.

I removed setx and the option to permanently activate from construct_env, but it's still used as a fallback to set registry variables where it already had the 1024 check in place.

I fixed the PostMessage error that was happening by using SendMessageWithTimeout instead. It only sends the message if a registry value actually changed. I also confirmed that the message is being received by other applications.

The one caveat that still exists is that you need to run emsdk activate before running emsdk activate --global so that it will use the correct python version.

### CODE REVIEW

**Comment by sbc100** [comment]
Regarding pywin32.  This should now be included in the version of python that is downloaded by emsdk.  At least as of #487..  can you confirm?

**Comment by sbc100** [comment]
Seems like a good change to me.   @juj?

File: emsdk.py
Comment (juj) [inline emsdk.py:2489]: Hmm, there are two issues with this:

1. `construct_env` is intended to create the `emsdk_env.bat` script that one can use to enter the Emsdk-enabled command prompt locally. It should never globally set environment variables.
2. Not a fault of this PR, but `construct_env` should actually remove the permanently_activate=true support. In current emsdk, it is always called with permanently_activate=False, to avoid a catastrophic bug where `%PROGRAMFILES%` etc. wildcards were getting expanded in PATH. (and also https://github.com/emscripten-core/emsdk/issues/4). The rule of thumb is that one should never be using the `SETX` command to set system PATH.

The right location to do this would be later in the code block starting from the comment
`# Apply environment variables to global all users section.`, looks like that is missing the relevant registrations.

**Comment by sbc100** [comment]
This change is a lot bigger now :(   Can it be simplified or broken down into smaller chunks?   I'm having trouble groking all the moving parts there.

File: emsdk.py
Comment (kudaba) [reply emsdk.py]: So, a big update here. While the code remains to be able to set user environment variables in windows it really only runs in two modes: current session locals or system wide environment variables. I removed the code that was trying to delete the variables and causing errors since it doesn't really do anything.

I removed setx and the option to permanently activate from construct_env, but it's still used as a fallback to set registry variables where it already had the 1024 check in place.

I fixed the PostMessage error that was happening by using SendMessageWithTimeout instead. It only sends the message if a registry value actually changed. I also confirmed that the message is being received by other applications.

The one caveat that still exists is that you need to run ``emsdk activate`` before running ``emsdk activate --global`` so that it will use the correct python version.

File: emsdk.py
Comment (sbc100) [inline emsdk.py:3026]: I also noticed this strange line.  I think we can maybe split this out as a separate change to simplify this one?   As far as I know this "perm" argument is not documented or used, right @juj?

Comment (kudaba) [reply emsdk.py:3026]: That and I removed the secondary argument for that function.

Comment (sbc100) [reply emsdk.py:3026]: I split that part out into separate change #525, in the hope that it will simplify this one.    If you'd rather do that yourself (to avoid re-basing on top of my change) please go ahead.

**Comment by sbc100** [comment]
If there are part of this change that are cleanups or re-structuring only perhaps you could submit those separately?

Also, perhaps the PR description could be update to reflect exactly what this change is supposed to do?  Or is it up-to-date already?

**Comment by kudaba** [comment]
Updated the PR with the info of all the changes. I could go through an restructure it, unfortunately I won't have the time to get back to this for a while.

**Comment by sbc100** [comment]
How does this relate to https://github.com/emscripten-core/emsdk/pull/468  ?

**Comment by kudaba** [comment]
There will be merge conflicts but they should be able to co-exists.

**Comment by aminya** [comment]
Can we get this rebased? This will probably fix the windows tests.

File: emsdk.py
Comment (sbc100) [inline emsdk.py:2596]: Is this added an extra empty line to the output?

File: emsdk.py
Comment (sbc100) [inline emsdk.py]: Doesn't flake8 complain about this if you don't have two lines between each function?

File: emsdk.py
Comment (sbc100) [inline emsdk.py:2560]: Just `if env_vars_to_add:` should be fine here.

File: emsdk.py
Comment (sbc100) [inline emsdk.py:2444]: Split the params onto separate lines to avoid the very long line here?

File: emsdk.py
Comment (sbc100) [inline emsdk.py:2427]: Just `if env_vars_to_add:` should be fine here.

File: emsdk.py
Comment (sbc100) [inline emsdk.py:2446]: Might be nice to keep this entire block as a seperate function.   Then you could avoid a lot of the nesting by having early returns after `if not changed:` and after `if env_vars_to_add:`?

File: emsdk.py
Comment (kudaba) [reply emsdk.py:2596]: yeah, just to make formatting a bit nicer

File: emsdk.py
Comment (kudaba) [reply emsdk.py]: I force pushed the formatting fixes from flake

**Review by aminya — APPROVED** [review]
I tested this. It works! 

I recommend enabling the tests:
https://github.com/kudaba/emsdk/blob/ca1e15fb6a66c4d62e2a72a27cbff371b4e2a09f/.circleci/config.yml#L95
```diff
          command: |
-            try {
               scripts/test_permanent.ps1
-            } catch {
-             Write-Host "`n Error Message: [$($_.Exception.Message)"] -ForegroundColor Red
-              Write-Host "`n Line Number: "$_.InvocationInfo.ScriptLineNumber -ForegroundColor Red
-            }
```

**Comment by aminya** [comment]
I am happy to say that I tested this in combination with #635, and all the `scripts/test_permanent.ps1` tests finally pass! 

We should definitely enable them as I showed in https://github.com/emscripten-core/emsdk/pull/515#pullrequestreview-503497305

After #635 is merged there is a small conflict that needs to be fixed. I removed the following lines in that PR, so after accepting your changes just remove this line:
```
prev_path = os.environ['PATH']
```

**Comment by kudaba** [comment]
Sorry for all the force pushes, it took me a while to get local testing working.

I fixed up the exception handling with the registry and had to fix up the permanent test script a bit as well since it assumed emsdk was being installed to a folder called emsdk.

I did notice that the registry handling is a lot slower now, but since it's generally a one time thing it's probably not anything worth worrying about.

**Comment by aminya** [comment]
Awesome! 

I did not notice any performance difference. Anyways, I think stability is more important. 😃

**Comment by aminya** [comment]
@sbc100 Can we roll this in?

File: emsdk.py
Comment (sbc100) [reply emsdk.py:2584]: But why do you say this is nicer?   Right no I see the following output on my machine:

```
sbc.sbc-glaptop (~/dev/wasm/emsdk) ((53c184e...)) $ . ./emsdk_env.sh 
Adding directories to PATH:
PATH += /home/sbc/dev/wasm/emsdk
PATH += /home/sbc/dev/wasm/emsdk/upstream/emscripten
PATH += /home/sbc/dev/wasm/emsdk/node/12.18.1_64bit/bin

Setting environment variables:
EMSDK = /home/sbc/dev/wasm/emsdk
EM_CONFIG = /home/sbc/dev/wasm/emsdk/.emscripten
EM_CACHE = /home/sbc/dev/wasm/emsdk/upstream/emscripten/cache
EMSDK_NODE = /home/sbc/dev/wasm/emsdk/node/12.18.1_64bit/bin/node
sbc.sbc-glaptop (~/dev/wasm/emsdk) ((53c184e...)) $
```

To me an extra empty line at the end would look pointless/messy.. but maybe that is just me?

**Review by sbc100 — COMMENTED** [review]
I think we still have the problem that we will clobber any local changes that the user has in their PATH with a fresh copy from the registry... but that is separate issue I think.

File: emsdk.py
Comment (sbc100) [inline emsdk.py:349]: I'd love to know under what circumstances these two excpt blocks can fire?   Do we really need two nested try blocks here?  Again, not really related to this PR.

**Review by sbc100 — APPROVED** [review]

File: emsdk.py
Comment (aminya) [reply emsdk.py:445]: @kudaba has added comments for these. The original code did not have any description
https://github.com/emscripten-core/emsdk/pull/515/files#diff-7ecf1f48f2952ff372972476842c8882L321

File: emsdk.py
Comment (aminya) [reply emsdk.py:338]: `winreg.OpenKey` is not "noexcept", so it may throw. The documentation says the type of error would be OSError.
https://docs.python.org/3/library/winreg.html#winreg.OpenKey

File: emsdk.py
Comment (sbc100) [reply emsdk.py:445]: I mean more specific in the type of exceptions we catch here?  Can it be just OSError perhaps?

File: emsdk.py
Comment (kudaba) [reply emsdk.py:349]: Oh, I can clean that up, I think it was there to try to catch exceptions from opening the registry key. I almost did the same in set variable but didn't notice that the get was already doubled up.

File: emsdk.py
Comment (aminya) [reply emsdk.py:445]: I don't it hurts to catch all the exceptions. We are not concerned with the performance of this program that runs on a powerful desktop processor. The stability is more important, IMO.

Comment (kudaba) [reply emsdk.py:442]: I'm not sure how error handling or exceptions are working with the interop between python and win32 so this is just me being a bit extra cautious.

**Comment by kudaba** [comment]
> I think we still have the problem that we will clobber any local changes that the user has in their PATH with a fresh copy from the registry... but that is separate issue I think.

Actually I think this will be fine. It adds emsdk paths to the local path environment variable and the registry setting separately so it should be additive in both cases. I had to do that since just setting the registry setting was not actually changing the currently set environment variable.

**Comment by sbc100** [comment]
> > I think we still have the problem that we will clobber any local changes that the user has in their PATH with a fresh copy from the registry... but that is separate issue I think.
> 
> Actually I think this will be fine. It adds emsdk paths to the local path environment variable and the registry setting separately so it should be additive in both cases. I had to do that since just setting the registry setting was not actually changing the currently set environment variable.

The problem is not about addings to the current env.   I believe the problem is that we always take the registry value as the *initial* PATH to add to (we've add the concept of fallback here but I think in practice it won't fire).   This means that you always end up with "registry value + extras" rather than "current shell value + extras".  

Basically we need to calculate two different paths which can be different:

1. What path should be added the registery.
2. What path should be set in the current shell

Again, its not something we need to fix as part of this PR, I guess its bug we have always had.

File: emsdk.py
Comment (sbc100) [reply emsdk.py:442]: Catching specific exceptions is not about performance but about code quality and maintainability.   In general we want to minimize the size of the try block and be as specific as possible in the exceptions we catch.  

From pep8:  `When catching exceptions, mention specific exceptions whenever possible instead of using a bare except: clause:`

One problem with catching everything is that it can hide all kinds of programmer errors. e.g. simple things like accessing out of bounds which the except block is almost never supposed to handle.

Comment (sbc100) [reply emsdk.py:442]: Anyway, unrelated to this PR so we don't need to block on this.

**Comment by sbc100** [comment]
I think this change is ready to land now.

Can you craft as suitable message for the squashed commit?    I would use the PR description but I'm not sure which parts are still relevant?

**Comment by kudaba** [comment]
The first commit has the most accurate messaging I think:

Fix environment variables to be consistently set when using --system or --permanent
Switch PostMessage with SendMessageWithTimeout and only send one if environment variables actually changed
Enable and fix permanent unit tests

Or did you mean you want me to squash the review feeback commit back down into the original (I had been doing it this whole time until the last bit of feedback).

**Comment by sbc100** [comment]
Thanks!

**Pull request merged**
